### PR TITLE
React Native Profiling Docs

### PR DESCRIPTION
**Pull request by MarcoPolo**
After diving into profiling our android app, I found the very act of profiling it confusing. It also seems like not many people play around with these things (i.e. I haven't see anyone else make use of RN's native systrace tools). Instead of going on a deeper dive into some of the possible fixes I found, I thought it would be more helpful to make a write up of profiling so others can also contribute ideas and suggestions as well as be empowered to do their own profiling tests.

The write up is pretty basic. It has a list of useful links, some tips, and a couple case studies.

I'm going to somewhat randomly nominate @buoyad to review this for coherency and see if he can do a simple profile of the app using Systrace to time how long it takes for us to startup to the people tab.

cc @keybase/react-hackers

### CODE REVIEW

File: shared/docs/react-native/profiling.md [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+## ReactNative.Systrace
+
+Api [here](https://facebook.github.io/react-native/docs/systrace)
+
+This is super useful in analyzing startup time performance cost. You can use
+this to see how long requires are taking. But there's a trick. You have to
+change the implementation of a couple methods for it to work properly.
+
+Add this to your entry file to enable Systrace (you can also enable it via RN
```
Comment (buoyad) [inline shared/docs/react-native/profiling.md]: I can't get this to work, `Can't find variable: performance` (I put it in `index.android`). More specifics for where this should be? Also, button path for enabling via RN devtools would be nice.

File: shared/docs/react-native/profiling.md [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+with the engine. Part of this is that we need to require so many files before
+we can start the work with the engine.
+
+I tried moving the setup that happens in `app/index.native` to happen before.
+But I didn't see a signifcant startup time difference. My guess is that without
+changing all files necessary to start the bootstrapping, we still load most of
+the app. So there isn't a real difference.
+
+
+### Areas to improve (in order or biggest impact and easiest to implement)
```
Comment (buoyad) [inline shared/docs/react-native/profiling.md]: in order of*

File: shared/docs/react-native/profiling.md [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+trace-event payload.
+
+[The simple node server](https://gist.github.com/MarcoPolo/0d3186becbce74bc8f21c9deab33dcc0)
+
+Run this a couple of times to get the starting number that you're going to try
+to improve.
+
+### Investigating slowdowns
+
+I enabled RN's systrace with the method listed above. as soon as
```
Comment (buoyad) [inline shared/docs/react-native/profiling.md]: How did you do this if (from above):
> This doesn't work outside of dev mode.

and
> gradle installReleaseUnsigned

File: shared/docs/react-native/profiling.md [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+* Investigate if we can get by with loading a minimal version of emoij data
+
+
+## Case study: Testing immediate engine calls during bootstrap.
+
+### Abstract
+
+Normally the engine call is setup and called with a `setImmediate`. This is
+usually exactly what we want. Trading a bit of response time from the engine
+for a more response ui. During bootstrapping, our priorities our a little
```
Comment (buoyad) [inline shared/docs/react-native/profiling.md]: our priorities are*

File: shared/docs/react-native/profiling.md [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+makes sense to avoid that tradeoff.
+
+We'll test if avoiding `setImmediate` during bootstrapping will help reduce
+startup time.
+
+### Test setup
+
+I cherry picked these commits to start the testing process:
+
+* `bd32cdee4c` - Android perf helper files
```
Comment (buoyad) [inline shared/docs/react-native/profiling.md]: Did you push these commits?
```
$ git cat-file -t bd32cdee4c
fatal: Not a valid object name bd32cdee4c
```

Comment (buoyad) [reply shared/docs/react-native/profiling.md]: (same for `507f` and `770c`)

File: shared/docs/react-native/profiling.md (hunk 1 above)
Comment (MarcoPolo) [reply shared/docs/react-native/profiling.md]: What do you mean button path?

Comment (MarcoPolo) [reply shared/docs/react-native/profiling.md]: I forgot to mention how to add the performance polyfill! which is needed if you aren't connected to a remote debugger.

File: shared/docs/react-native/profiling.md (hunk 5 above)
Comment (MarcoPolo) [reply shared/docs/react-native/profiling.md]: whoops! You're right, I just pushed it and swapped these with links

File: shared/docs/react-native/profiling.md (hunk 1 above)
Comment (buoyad) [reply shared/docs/react-native/profiling.md]: Ah, that'd explain it. By button path I mean I don't know how to enable it via RN devtools.

Comment (MarcoPolo) [reply shared/docs/react-native/profiling.md]: Ah yeah, I never used it, but it's there: 



I'll remove that sentence

Comment (buoyad) [reply shared/docs/react-native/profiling.md]: Ah that option isn't on android

**Comment by MarcoPolo** [comment]
Thanks @buoyad!!!

**Pull request merged**